c++/src/kj/memory.h

```diff
@@ -168,6 +170,53 @@ class NullDisposer: public Disposer {
   void disposeImpl(void* pointer) const override {}
 };
 
+
+// =======================================================================================
+// Ptr Counters
+
+#ifdef KJ_DEBUG
+#define KJ_ASSERT_PTR_COUNTERS
+// When defined, keeps track of active Ptr<T> instances and asserts validity of their ownership
+#endif
+
+namespace _ {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+class AtomicPtrCounter {
+// AtomicPtrCounter uses atomic operations to keep track of active pointers.
+// 
+// Thread safety notes:
+// - every dec() observes the effect of corresponding inc() because of release/consume order
+// - when assertEmpty() is called with balanced dec/inc calls, all releases were consumed, and
+//   additional load with consume memory order will see the correct last value.
+// - when assertEmpty() is called with unbalanced dec/inc calls int will observe the result of 
+//   one of the unbalanced inc(), which is always > 0.
```

Comment:
```suggestion
  // AtomicPtrCounter uses atomic operations to keep track of active pointers.
  // 
  // Thread safety notes:
  // - every dec() observes the effect of corresponding inc() because of release/consume order
  // - when assertEmpty() is called with balanced dec/inc calls, all releases were consumed, and
  //   additional load with consume memory order will see the correct last value.
  // - when assertEmpty() is called with unbalanced dec/inc calls int will observe the result of 
  //   one of the unbalanced inc(), which is always > 0.
```